mmv1/third_party/terraform/services/apigee/resource_apigee_api.go

```diff
@@ -0,0 +1,403 @@
+					if config_bundle, ok := d.GetOkExists("config_bundle"); ok {
+						localMd5Hash = tpgresource.GetFileMd5Hash(config_bundle.(string))
+					}
+					if localMd5Hash == "" {
+						return false
+					}
+
+					// `old` is the md5 hash we speculated from server responses,
+					// when apply responded with succeed, hash is set to the hash of uploaded bundle
+					if old != localMd5Hash {
+						return false
+					}
+
+					return true
+				},
+			},
+		},
+		UseJSONNumber: true,
+	}
+}
+
+func resourceApigeeApiCreate(d *schema.ResourceData, meta interface{}) error {
+	ctx := context.TODO()
+	tflog.Info(ctx, "resourceApigeeApiCreate")
+	log.Printf("[DEBUG] resourceApigeeApiCreate")
+
+	log.Printf("[DEBUG] resourceApigeeApiCreate, name=			 	%s", d.Get("name").(string))
+	log.Printf("[DEBUG] resourceApigeeApiCreate, org_id=, 			%s", d.Get("org_id").(string))
+	log.Printf("[DEBUG] resourceApigeeApiCreate, config_bundle=, 	%s", d.Get("config_bundle").(string))
+
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+
+	var file *os.File
+	var localMd5Hash string
+	if configBundlePath, ok := d.GetOk("config_bundle"); ok {
+		var err error
+		file, err = os.Open(configBundlePath.(string))
+		if err != nil {
+			return err
+		}
+		localMd5Hash = tpgresource.GetFileMd5Hash(configBundlePath.(string))
+	} else {
+		return fmt.Errorf("Error, \"config_bundle\" must be specified")
+	}
+
+	url, err := tpgresource.ReplaceVars(d, config, "{{ApigeeBasePath}}organizations/{{org_id}}/apis?name={{name}}&action=import")
+	if err != nil {
+		return err
+	}
+	billingProject := ""
+
+	// err == nil indicates that the billing_project value was found
+	if bp, err := tpgresource.GetBillingProject(d, config); err == nil {
+		billingProject = bp
+	}
+
+	log.Printf("[DEBUG] resourceApigeeApiCreate, url=, 	%s", url)
+	res, err := sendRequestRawBodyWithTimeout(config, "POST", billingProject, url, userAgent, file, "application/octet-stream", d.Timeout(schema.TimeoutCreate))
+
+	log.Printf("[DEBUG] sendRequestRawBodyWithTimeout Done")
+	if err != nil {
+		return fmt.Errorf("Error creating API proxy: %s", err)
+	}
+
+	// Store the ID now
+	id, err := tpgresource.ReplaceVars(d, config, "organizations/{{org_id}}/apis/{{name}}")
+	if err != nil {
+		return fmt.Errorf("Error constructing id: %s", err)
+	}
+	d.SetId(id)
+	log.Printf("[DEBUG] create d.SetId done, id = %s", id)
+
+	log.Printf("[DEBUG] Finished creating API proxy %q: %#v", d.Id(), res)
+
+	if resourceApigeeApiRead(d, meta) != nil {
+		return fmt.Errorf("Error reading API proxy at end of Create: %s", err)
+	}
+	d.Set("md5hash", localMd5Hash)
+	d.Set("detect_md5hash", localMd5Hash)
+	return nil
+}
+
+func resourceApigeeApiUpdate(d *schema.ResourceData, meta interface{}) error {
+	//For how API proxy api is implemented, just treat an update as create, when the name is same, it will create a new revision
```

Comment:
If calling the create method creates a new revision, how do we make sure we are reading the right revision when we call the read method? If someone creates a new revision of an API outside of Terraform would we immediately see a diff?

Reply:
Yes, that should trigger a diff and that is expected the proxy would have changed outside the local copy is different, so the expectation would be to have to re-apply to bring back the version locally.